<a href="https://colab.research.google.com/github/EmamulHossen/KNN_Assignment_7.2/blob/main/KNN_for_Regression(7_2).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive

In [ ]:
drive.mount("/content/drive")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# **Import Required Libraies**

In [ ]:
import pandas as pd
import numpy as np

**1.Import the dataset**

In [ ]:
df=pd.read_csv("/content/drive/MyDrive/Data/weight-height.csv")

In [ ]:
df.head()

Gender     Height      Weight
0   Male  73.847017  241.893563
1   Male  68.781904  162.310473
2   Male  74.110105  212.740856
3   Male  71.730978  220.042470
4   Male  69.881796  206.349801

In [ ]:
df.Gender = df['Gender'].replace(['Male', 'Female'], [1,0])
df.head()

Gender     Height      Weight
0       1  73.847017  241.893563
1       1  68.781904  162.310473
2       1  74.110105  212.740856
3       1  71.730978  220.042470
4       1  69.881796  206.349801

In [ ]:
df.tail()

Gender     Height      Weight
8550       0  60.483946  110.565497
8551       0  63.423372  129.921671
8552       0  65.584057  155.942671
8553       0  67.429971  151.678405
8554       0  60.921791  131.253738

In [ ]:
#Shape of Dataset
print("Shape of the dataset: {}".format(df.shape))
print("Shape of dataset: Number of Row:{},Number of Column:{}".format(df.shape[0],df.shape[1]))

Shape of the dataset: (8555, 3)
Shape of dataset: Number of Row:8555,Number of Column:3


**Separate X(Gender,Height) and Y (Weight)**

In [ ]:
x=df.drop(columns=["Weight"], axis=1)

In [ ]:
x.head()

Gender     Height
0       1  73.847017
1       1  68.781904
2       1  74.110105
3       1  71.730978
4       1  69.881796

In [ ]:
y=df["Weight"]

In [ ]:
y.head()

0    241.893563
1    162.310473
2    212.740856
3    220.042470
4    206.349801
Name: Weight, dtype: float64

**Split the dataset as Train=70% & Test=30%**

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.3,random_state=42)

In [ ]:
print("Shape of x_train:",x_train.shape)
print("Shape of x_test :",x_test.shape)
print("Shape of y_train",y_train.shape)
print("Shape of y_test:",y_test.shape)

Shape of x_train: (5988, 2)
Shape of x_test : (2567, 2)
Shape of y_train (5988,)
Shape of y_test: (2567,)


# **Apply Linear Regression**

In [ ]:
from sklearn.linear_model import LinearRegression


In [ ]:
#Creation Object
model=LinearRegression()

In [ ]:
model.fit(x_train,y_train)

LinearRegression()

In [ ]:
model.coef_

array([19.54151693,  5.9562333 ])

In [ ]:
model.intercept_

-243.79306041624142

In [ ]:
#Prediction
y_predic_train=model.predict(x_train)

In [ ]:
y_predic_train

array([178.35427565, 202.22486558, 129.22191775, ..., 139.34406368,
       187.17224616, 158.78183812])

In [ ]:
y_train

553     186.751417
1397    211.031652
7934    143.768451
8367    151.814648
3320    196.505814
           ...    
5734    139.223151
5191     90.728561
5390    156.800504
860     179.386817
7270    150.380597
Name: Weight, Length: 5988, dtype: float64

In [ ]:
y_predic_test=model.predict(x_test)

In [ ]:
y_predic_test

array([142.44868114, 186.0517281 , 198.42283706, ..., 102.94080643,
       190.83713617, 143.34786753])

In [ ]:
y_test

6006    149.668369
1197    197.642244
2862    178.551191
6497    141.343095
2860    195.322675
           ...    
184     202.046425
2988    165.955953
6048     98.869253
420     190.933401
5669    123.868429
Name: Weight, Length: 2567, dtype: float64

**Performance**

In [ ]:
training_accuracy=model.score(x_train,y_train)

In [ ]:
training_accuracy

0.8973793060969246

In [ ]:
from sklearn.metrics import r2_score

In [ ]:
train_score=r2_score(y_train,y_predic_train)

In [ ]:
train_score

0.8973793060969246

In [ ]:
test_score=r2_score(y_test,y_predic_test)

In [ ]:
test_score

0.905911242442266

In [ ]:

testing_accuracy=model.score(x_test,y_test)

In [ ]:
testing_accuracy

0.905911242442266

In [ ]:
from sklearn.metrics import mean_squared_error
MSE = mean_squared_error(y_test,y_predic_test)

In [ ]:
MSE

96.8373443783061

# **Apply KNN Regressor**

In [ ]:
 from sklearn.neighbors import KNeighborsRegressor

In [ ]:
kNN_model=KNeighborsRegressor(n_neighbors=3)

In [ ]:
kNN_model.fit(x_train,y_train)

KNeighborsRegressor(n_neighbors=3)

In [ ]:
#Prediction
y_train_prediction=kNN_model.predict(x_train)

In [ ]:
y_train_prediction

array([178.3390928 , 209.35667097, 132.54853953, ..., 144.26695503,
       194.6310546 , 159.7512038 ])

In [ ]:
y_train

553     186.751417
1397    211.031652
7934    143.768451
8367    151.814648
3320    196.505814
           ...    
5734    139.223151
5191     90.728561
5390    156.800504
860     179.386817
7270    150.380597
Name: Weight, Length: 5988, dtype: float64

In [ ]:
y_test_prediction=kNN_model.predict(x_test)

In [ ]:
y_test_prediction

array([142.5130416 , 173.90630823, 184.6686197 , ...,  95.44597987,
       182.99507227, 147.37662317])

**KNN model performance**

In [ ]:
#Train Accuracy
train_accuracy=kNN_model.score(x_train,y_train)

In [ ]:
train_accuracy

0.9304521916751347

In [ ]:
#Test Accuracy
test_accuracy=kNN_model.score(x_test,y_test)

In [ ]:
test_accuracy

0.8679879688589832

In [ ]:
from sklearn.metrics import mean_squared_error
KNN_MSE = mean_squared_error(y_test,y_test_prediction)

In [ ]:
KNN_MSE

135.8684592453893

# **Compare these two model**
>  **For Linear regression Model:**<br/>
      1.**Training Accuracy 0.8973793060969246**<br/>
      2.**Testing accuracy 0.905911242442266**<br/>
      **Here testing accuracy is greater than training accuracy .So we can say that the model is overfiting**

>  **For KNN regressor Model:**<br/>
      1.**Traing accuracy 0.9304521916751347**<br/>
      2.**Testing accuracy 0.8679879688589832**<br/>
      **Testing accuracy is less than training accuracy**

# **The KNN Regressor model is better than Linear Regression for this dataset.**




